pkg/ociwclayer/export.go

```diff
@@ -76,6 +78,27 @@ func writeTarFromLayer(ctx context.Context, r hcsshim.LayerReader, w io.Writer)
 				return err
 			}
 		} else {
+			numberOfLinks, fileIDInfo, err := r.LinkInfo()
+			if err != nil {
+				return err
+			}
+			if numberOfLinks > 1 {
+				if linkName, ok := linkRecords[fileIDInfo.FileID]; ok {
+					// We've seen this file before, by another name, so put a hardlink in the tar stream.
+					hdr := backuptar.BasicInfoHeader(name, 0, fileInfo)
+					hdr.Mode = 0644
+					hdr.Typeflag = tar.TypeLink
+					hdr.Linkname = linkName
+					if err := t.WriteHeader(hdr); err != nil {
+						return err
+					}
+					continue
+				}
+
+				// All subsequent names for this file will be hard-linked to this name
+				linkRecords[fileIDInfo.FileID] = filepath.ToSlash(name)
+			}
+
```

Comment:
Was there a specific need for this block or was this an optimization?

Reply:
The whole block? This is the code to generate hard-links in the tar stream from hard-links seen on-disk. As a practical observation, the UtilityVM is mostly hard-links (some are also created by the image unpacking), so without this, unpacking and repacking the nanoserver image I was testing against doubled the size of the tarball.

Reply:
See 811c3fc1fb75ea3e2ecaf2e3e3b26e3d71ebe964 specifically. This is the commit I was referring to that pulled in the containerd and go-winio upgrades, if we want to put it aside for later to unblock the rest of this PR.

Reply:
Ah, sorry, had missed that. I think it's fine to keep in this PR.